File: 2025/October/7/bintreelevordertravel2.js
// LeetCode 107: Binary Tree Level Order Traversal II
// Explanation:
// 1. Use BFS with a queue.
// 2. Add each level’s nodes to result.
// 3. Reverse result at the end for bottom-up order.
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
var levelOrderBottom = function(root) {
    if (!root) return [];
    let queue = [root], res = [];
    while (queue.length) {
        let level = [];
        let size = queue.length;
        for (let i = 0; i < size; i++) {
            let node = queue.shift();
            level.push(node.val);
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
        res.unshift(level);
    }
    return res;
};

let root = { val: 3, left: { val: 9 }, right: { val: 20, left: { val: 15 }, right: { val: 7 } } };
console.log(levelOrderBottom(root)); // [[15,7],[9,20],[3]]

let root2 = { val: 1 };
console.log(levelOrderBottom(root2)); // [[1]]

console.log(levelOrderBottom(null)); // []

File: 2025/October/7/bintreelevordertravel2.py
# LeetCode 107: Binary Tree Level Order Traversal II
# Explanation:
# 1. Perform a standard BFS level order traversal.
# 2. Collect nodes level by level.
# 3. Reverse the list of levels at the end to get bottom-up order.
# Time Complexity: O(n) — each node is visited once.
# Space Complexity: O(n) — queue and result storage.

from collections import deque
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        queue = deque([root])
        res = []
        while queue:
            level = []
            for _ in range(len(queue)):
                node = queue.popleft()
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            res.append(level)
        return res[::-1]

root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20, TreeNode(15), TreeNode(7))
print(Solution().levelOrderBottom(root))  # [[15, 7], [9, 20], [3]]

root = TreeNode(1)
print(Solution().levelOrderBottom(root))  # [[1]]

print(Solution().levelOrderBottom(None))  # []
